contracts/eosio.system/src/peer_keys.cpp

```diff
@@ -47,15 +46,57 @@ peer_keys::getpeerkeys_res_t peer_keys::getpeerkeys() {
    getpeerkeys_res_t resp;
    resp.reserve(max_return);
 
-   auto idx = producers.get_index<"prototalvote"_n>();
+   double vote_threshold = 0; // vote_threshold will always be >= 0
 
-   for( auto it = idx.cbegin(); it != idx.cend() && resp.size() < max_return && it->total_votes > 0 && it->active(); ++it ) {
+   auto add_peer = [&](auto it) {
       auto peers_itr = peer_keys_table.find(it->owner.value);
       if (peers_itr == peer_keys_table.end())
          resp.push_back(peerkeys_t{it->owner, {}});
       else
          resp.push_back(peerkeys_t{it->owner, peers_itr->get_public_key()});
-   }
+
+      // once 21 producers have been selected, we will only consider producers that have more
+      // than 50% of the votes of the 21st selected producer.
+      // ------------------------------------------------------------------------------------
+      if (resp.size() == 21)
+         vote_threshold = std::abs(it->total_votes) * 0.5;
+   };
+
+   auto idx = producers.get_index<"prototalvote"_n>();
+
+   auto it  = idx.cbegin();
+   auto rit = idx.cend();
+   if (it == rit)
+      return resp;
+   else
+      --rit;
+
+   // 1. Consider both active and non-active producers. as a non-active producer can be
+   //    reactivated at any time.
+   // 2. Once we have selected 21 producers, the threshold of votes required to be selected
+   //    increases from `> 0` to `> 50% of the votes that the 21st selected producer has`.
+   // 3. We iterate from both ends as non-active producers have their vote total negated, so
+   //    the highest voted non-active producer will be the last entry of our index.
+   // --------------------------------------------------------------------------------------
+   bool last_one = false;
+   do  {
+      // invariants:
+      //   - `it` and `rit` both point to a valid `producer_info` (possibly the same)
+      //   - `it` <= `rit`
+      // ----------------------------------------------------------------------------
+      last_one = (it == rit);
+      if (rit->total_votes < vote_threshold && (-rit->total_votes > std::abs(it->total_votes))) {
+         add_peer(rit);
+         if (!last_one)
+            --rit;
+      } else if (it->total_votes > vote_threshold) {
+         add_peer(it);
+         ++it;
+      } else {
+         break;
+      }
+   } while (!last_one && resp.size() < max_return);
```

Comment:
I think this loop could instead be rewritten to be something like the following:

```cpp
while (resp.size() < max_return) {
   if (vote_threshold < rit->total_votes && it->total_votes < rit->total_votes) {
      add_peer(rit);
      assert(it != rit); // Should always be satisfied since it->total_votes != rit->total_votes
      --rit;
      continue;
   }
   if (vote_threshold < it->total_votes) {
      add_peer(it);
      if (it == rit) 
         break;
      ++it;
      continue;
   }
   break;
}
```

Reply:
You'd have to use `std::abs(rit->total_votes)`  if both occurences in the first if condition to consider any negative `total_votes`, otherwise suppose you have this in your index:
```
-5
-15
```

would you return anything?

If we do that, though, the justification on the assert does not hold anymore.

Reply:
Do you like it better like that?